Add the digest-environment dependency configuration

abderrahim opened the pull request:
This allows setting an environment variable inside the sandbox to the CAS digest of one or more dependencies.

This us useful when communicating with the nested remote execution endpoint, but could also be used independently.

This PR replaced #1991.

abderrahim committed:
buildelement.py: Add the digest-environment config property

This allows setting an environment variable inside the sandbox to the CAS
digest of one or more dependencies.

Co-authored by: Adrien Plazas <adrien.plazas@codethink.co.uk>

abderrahim committed:
sandbox.py: Add method to create a sub-sandbox

This allows an element to use a secondary sandbox for manipulating
artifacts that don't affect the build.

abderrahim committed:
_sandboxremote.py: Upload blobs for subsandbox roots

The subsandboxes can be used to extract a CAS digest that could be used for
nested remote execution, and thus need to be available in the remote execution
CAS.

abderrahim commented on src/buildstream/buildelement.py:
```diff
@@ -285,10 +305,20 @@ def configure_sandbox(self, sandbox):
-        sandbox.set_environment(self.get_environment())
-
     def stage(self, sandbox):
+        # Setup environment
```
I've moved this here because buildstream wouldn't let me stage outside of `stage()`.

It might be better to try to work around that limitation rather than doing this virtual staging here.

abderrahim replied:
This is due to the overlap collector, btw. The staging functions will refuse to work if there is no active overlap collector.

juergbi replied:
Setting up the environment in `stage()` is a problem at least for `bst shell --build --use-buildtree`, which does not invoke `Element.stage()` but still requires environment variables to be set.

Draft fix for the overlap collector issue and moving environment setup back to `configure_sandbox()`: https://github.com/apache/buildstream/commits/jbilleter/digest-environment/

juergbi replied:
I've pushed the fix to this branch and it seems to work well for normal builds. However, it still doesn't work for `bst shell --build --use-buildtree` (see failing test case) as that skips the loading of dependency artifacts. Always requiring build dependencies to be available for `bst shell --build --use-buildtree` would fix the test case but I suspect it would break buildtree shells for artifacts. Any suggestions?

gtristan replied:
It looks to me that we do actually cache the environment variables in the artifact metadata, it seems to me that we should be relying on that in `bst shell --build --use-buildtree`.

juergbi replied:
Right, it's part of the low diversity metadata. I'll have to check whether this would allow us to completely skip `configure_sandbox()` for the buildtree shell.

However, even if that works, there is still one potential issue, which is that this would not guarantee that the directory tree referenced by the digest environment variable is available in the local cache, as the artifact doesn't reference that digest (except via that environment variable but BuildStream core is not aware of that). Maybe we can add that to the Artifact proto as well, but this may require additional code and API in the core for use by `BuildElement`. Or do you have another suggestion?

gtristan replied:
> However, even if that works, there is still one potential issue, which is that this would not guarantee that the directory tree referenced by the digest environment variable is available in the local cache, as the artifact doesn't reference that digest (except via that environment variable but BuildStream core is not aware of that). Maybe we can add that to the Artifact proto as well, but this may require additional code and API in the core for use by BuildElement. Or do you have another suggestion?

We are interestingly dancing on the line of sanity here.

* The assumption with `bst shell --build --use-buildtree` is that you have everything you need to reproduce the (probably failed) build locally in order to debug it.
* The assumption with exporting the local casd socket to the build environment, and moreso, remoting out the builds to a different sandbox (whether it is on the same machine or not)... is that you have your hands on some sensitive things, you know what you are doing, and you risk corrupting your build and cache with some non-determinism

The above two are not necessarily at odds with eachother, but it could be fair to say that if you are doing this wild stuff, you are out of bounds for having support for debugging remotely failed builds.

I think that I would be satisfied with either of the following:

* The officially supported and documented usage of `recc` is to use it locally in your sandbox, exercise the local compiler in the same sandbox, and use only the caching nature of `recc` to store/lookup objects in the CAS
  * This would "just work" given that we use the same environment variables from cached artifacts
* The officially supported and documented usage of `recc` is to selectively stage the compiler, source code and required header files to an adjacent sandbox on the same buildhost via the local `buildbox-casd` feature, and the artifact proto is extended to capture a list of digests which might be used for weird features like this one

I.e., lets support something, and if you are doing something outside of what is supported, you might be on your own and your build shells for debugging remotely failed artifacts might not work.

juergbi replied:
The intention is to recommend the latter, i.e., use the remote execution protocol even for local execution, as the former (cache-only) is much easier to get wrong and get incorrect cache hits due to incomplete cache keys.

juergbi replied:
I've implemented the Artifact proto enhancement and all tests are now passing locally.

juergbi replied:
Working on additional tests, I've noticed that `bst artifact checkout` may be broken for elements that use `digest-environment`. Need to look into this further.

juergbi replied:
I've added the commit `element.py: Don't call configure_sandbox() for non-build sandboxes` and expanded the tests to cover `bst artifact checkout`.

abderrahim commented on src/buildstream/buildelement.py:
```diff
@@ -285,10 +305,20 @@ def configure_sandbox(self, sandbox):
+            subsandbox = sandbox.create_subsandbox()
+            self.stage_dependency_artifacts(subsandbox, element_list)
+            digest = subsandbox.get_virtual_directory()._get_digest()
+            env[digest_variable] = "{}/{}".format(digest.hash, digest.size_bytes)
```
I also noticed that these environment variables don't end up in the build logs.

juergbi replied:
I've pushed an improvement to build environment logging.

abderrahim commented on src/buildstream/sandbox/_sandboxremote.py:
```diff
@@ -263,9 +263,17 @@ def _execute_action(self, action, flags):
+                # Add virtual directories for subsandboxes
+                for subsandbox in self._get_subsandboxes():
+                    vdir = subsandbox.get_virtual_directory()
+                    root_digests.append(vdir._get_digest())
```
I wonder if we should have an option to only upload these conditionally. It think it's not worth the trouble, but wanted to have a second opinion.

juergbi replied:
I also think it's not worth it. The files in subsandboxes should typically be a subset of the main input root, so the overhead should just be a few `Directory` protobuf messages.

abderrahim commented on src/buildstream/buildelement.py:
```diff
@@ -285,10 +305,20 @@ def configure_sandbox(self, sandbox):
+        for digest_variable in sorted_envs:
+            element_list = [element for element, _ in self.__digest_environment[digest_variable]]
+            subsandbox = sandbox.create_subsandbox()
+            self.stage_dependency_artifacts(subsandbox, element_list)
```
@gtristan was worried that the CAS digest might depend on the order the elements are declared in the .bst file

See https://github.com/apache/buildstream/pull/1991#discussion_r2153808770

juergbi replied:
The loader takes care of sorting dependencies via `loadelement.sort_dependencies()` early on, so `Element` and its subclasses never even have access to the declaration order and thus, there is no such issue, as far as I can tell.

gtristan replied:
While that is true, it is unclear to me whether the `Element` constructor interprets the `config:` of each dependency in dependency order, or in declaration order.

juergbi replied:
`Element._new_from_load_element()` iterates over the already sorted dependencies, creates a `DependencyConfiguration` object for each dependency with configuration data, and then hands that to `configure_dependencies()`, still in the same order. The declaration order can't even be observed by `Element`, as far as I can tell.

Does this clarify things or am I misunderstanding your comment?

gtristan replied:
Yes this clarifies things, this was precisely my concern :)

juergbi committed:
element.py: Support per-sandbox overlap collector

With the introduction of subsandboxes, a single overlap collector is no
longer sufficient.

juergbi committed:
element.py: Add `subsandbox()` method

juergbi committed:
element.py: Use sandbox environment for logging and shell

Elements can configure additional environment variables in
`configure_sandbox()`, which will be used by `BuildElement` to support
`digest-environment`. Use the complete environment for logging and
`bst shell`.

juergbi committed:
tests/integration: Add tests for `digest-environment`

juergbi committed:
Store sandbox state in Artifact proto

This is stored only when also storing the build tree. It is used to
configure the sandbox for `bst shell --use-buildtree`.

This is required for elements that create subsandboxes as part of
`configure_sandbox()`.

juergbi committed:
artifact.proto: Add a field for the build sandbox state

juergbi committed:
buildelement.py: Document `digest-environment`

abderrahim commented on src/buildstream/_artifactcache.py:
```diff
@@ -361,6 +361,9 @@ def _push_artifact_proto(self, element, artifact, artifact_digest, remote):
         if artifact_proto.buildroot:
             referenced_directories.append(artifact_proto.buildroot)
+        if artifact_proto.buildsandbox:
+            for subsandbox_digest in artifact_proto.buildsandbox.subsandbox_digests:
```
I think we need to push these in `_push_artifact_blobs()` as well.

juergbi replied:
Good catch. I've added that incl. test coverage.

abderrahim commented on src/buildstream/_artifact.py:
```diff
@@ -681,6 +696,21 @@ def pull(self, *, pull_buildtrees):
+    def configure_sandbox(self, sandbox):
+        artifact = self._get_proto()
+
+        if artifact.buildsandbox and artifact.buildsandbox.environment:
+            env = {}
+            for env_var in artifact.buildsandbox.environment:
+                env[env_var.name] = env_var.value
+        else:
+            env = self.load_environment()
```
I am not sure what this brings us. Why are we storing two different sets of environment variables in the artifact? Are they supposed to be different?

A bit of digging later: It seems the only relevance of the "static" environment is to compute the cache key, but we don't recompute the cache key from a loaded artifact. Am I missing something?

juergbi replied:
Loading the (new) sandbox environment in `load_environment()` would be wrong as it is used to set `Element.__environment`, which should not contain e.g. digest environment variables. It's possible that `Element.__environment` loaded from the artifact is not actually used right now, however, it may be used in the future e.g. by `bst inspect`. I didn't want to change anything in this regard to keep the risk of breakage or compatibility problems to a minimum.

abderrahim replied:
Alright, I don't have a strong opinion either way. Happy to keep this if you believe it reduces risk.

abderrahim commented on src/buildstream/buildelement.py:
```diff
@@ -286,10 +306,20 @@ def configure_sandbox(self, sandbox):
-        sandbox.set_environment(self.get_environment())
+        env = self.get_environment()
 
-    def stage(self, sandbox):
```
nit: removing the empty line after `def stage(self, sandbox):` results in a confusing diff.

abderrahim commented on tests/integration/project/elements/digest-environment/base-buildtree.bst:
```diff
@@ -0,0 +1,12 @@
+kind: manual
```
This file is identical to `base.bst`, any reason to have a separate element?

juergbi replied:
If I remember correctly, this is required for isolation with parallel runs of integration tests with the shared integration cache directory.

abderrahim replied:
Ack, I suspected as much. But I saw that there are two tests that share the same `base.bst` element. I guess the shared cache doesn't make a difference in that case?

juergbi replied:
Yes, `test_build_shell` doesn't actually build `base.bst`, so a reused cachedir shouldn't cause any issues.

abderrahim reviewed (commented):
A few comments but otherwise looks good.

(I can't approve on Github because I created the pull request)

abderrahim commented on src/buildstream/buildelement.py:
```diff
@@ -286,7 +306,18 @@ def configure_sandbox(self, sandbox):
+        for digest_variable in sorted_envs:
+            element_list = [element for element, _ in self.__digest_environment[digest_variable]]
+            with self.subsandbox(sandbox) as subsandbox:
+                self.stage_dependency_artifacts(subsandbox, element_list)
```
We should probably put this inside a timed activity with `silent_nested=True`. Logging is a bit too verbose (and confusing) right now.

juergbi replied:
Good point. I've updated the branch for this.

abderrahim commented on src/buildstream/element.py:
```diff
@@ -1453,7 +1453,11 @@ def _prepare_sandbox(self, scope, shell=False, integrate=True, usebuildtree=Fals
+
+                # Don't call `configure_sandbox()` as that may attempt to construct subsandboxes
+                # with build dependencies and we're not setting up a build sandbox.
+                sandbox.set_environment(self.get_environment())
```
I think it makes sense to squash this commit into the one before it.

abderrahim commented on src/buildstream/buildelement.py:
```diff
@@ -268,6 +278,16 @@ def get_unique_key(self):
             }
             dictionary["layout"] = layout_key
 
+        # Specify the layout in the key, if buildstream is to generate an environmenta
```
typo: "environmenta"

abderrahim reviewed (commented):
LGTM (with a couple nits)
